zfec/fec.c

```diff
@@ -394,11 +394,14 @@ _invert_vdm (gf* src, unsigned k) {
 }
 
 static int fec_initialized = 0;
-static void
-init_fec (void) {
-    generate_gf();
-    _init_mul_table();
-    fec_initialized = 1;
+
+void
+fec_init (void) {
```

Comment:
Thread synchronization isn't immediate. If thread 1 and 2 are both started, and thread 1 calls init, thread 2 won't necessarily see updated versions of the memory modified by thread 1, and even if it does it may be inconsistent.

The solution is to make `fec_initialized` an atomic, which is supported by modern C, and then read/write it with acquire/release atomic ordering.

https://marabos.nl/atomics/ is good book about this; Rust's model is based on C++'s model, and C's model is probably also C++'s model, so the semantics will all be the same even if the APIs are slightly different.

Reply:
Another alternative is to just create a static lock that needs to be acquired to initialized or check initialization.(Acquiring/releasing a lock is equivalent to acquire/release semantic on an atomic variable, so it too creates a happens-before relationship.)

This would simplify the API in that you don't need to make `fec_init()` public, you could just always call it. It would also make `fec_new()` slightly slower, but probably not in a very meaningful way.

Reply:
Thanks, great points here.  From some simple experiments, it looks like neither C11 "atomics" nor "threads" are available from the MSVC we get from current CI (eg, `stdatomic.h` exists but merely including it leads to numerous errors).  I don't quite understand this since we're using the Windows 2022 image and https://github.com/actions/runner-images/blob/main/images/win/Windows2022-Readme.md says this includes MSVC 17.7.  The build failures have a version number that looks like "14.35" ... but maybe that's not the MSVC version?  Maybe https://reactos.org/wiki/Visual_Studio_Versions is suggesting that "14.35" actually means "17.7".  I don't know.

So for now I'm going with "documentation".  Hopefully later someone can figure out how to get at atomics/threads on Windows and we can strengthen the implementation.

Reply:
Not that I

Reply:
A statically-initializable mutex here would work, yes.  There isn't one in C99 so we'd need a third-party library.  There is probably one but I am reluctant to give zfec it's first third-party build-time dependency.  It seems likely that pthreads offers some kind of solution here, which would at least solve the problem for Linux (maybe macOS too?), but it's not the obvious pthread_mutex_t because those are not statically initializable (so all you do is move the initialization problem from "coordinate zfec initialization" to "coordinate zfec initialization mutex initialization").  Windows probably has its own solution but I don't know what it is offhand either (there's at least 5 solutions that aren't statically initializable, for sure).

Reply:
I have vague memory that certain Python releases require certain MSVC versions? So older Python might get older MSVC or something.